sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,376 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_gpio.h"
+
+#include "gpio_regs.h"  // Generated.
+#include "sw/device/lib/common.h"
+
+/**
+ * Gives the mask that corresponds to the given bit index.
+ *
+ * @param index Bit index in a 32-bit register.
+ */
+static uint32_t index_to_mask(uint32_t index) { return 1u << index; }
+
+/**
+ * Perform a masked write to a GPIO register.
+ *
+ * The GPIO device provides masked bit-level atomic writes to its DIRECT_OUT
+ * and DIRECT_OE registers. This allows software to modify half of the bits
+ * at a time without requiring a read-modify-write. Note that depending on the
+ * value of the |mask|, this function may perform two writes.
+ *
+ * For instance, DIRECT_OUT's lower and upper halves can be modified by
+ * MASKED_OUT_LOWER and MASKED_OUT_UPPER, respectively. Upper half of
+ * MASKED_OUT_LOWER is the mask that determines which bits in the lower half of
+ * DIRECT_OUT will be modified, while lower half of MASKED_OUT_LOWER determines
+ * their values. MASKED_OUT_UPPER behaves in the same way for modifying the
+ * upper half of DIRECT_OUT.
+ *
+ * @param gpio GPIO instance.
+ * @param reg_lower_offset Offset of the masked access register that corresponds to
+ * the lower half of the actual register.
+ * @param reg_upper_offset Offset of the masked access register that corresponds to
+ * the upper half of the actual register.
+ * @param mask Mask that identifies the bits to write to.
+ * @param val Value to write.
+ */
+static void gpio_masked_write(const dif_gpio_t *gpio, uint32_t reg_lower_offset,
+                              uint32_t reg_upper_offset, uint32_t mask,
+                              uint32_t val) {
+  const uint32_t mask_lower_half = mask & 0x0000FFFFu;
```

Comment:
I don't know that you actually need the `u` when performing bitwise operations (I think the only time you need the u is for stuff like `1u << 32` to avoid shifting into the sign bit).

Maybe we should note this in the style guide...

Reply:
Yes, not needed here but I prefer using `u` on literals liberally just to be safe.